instance/tasks.py

```diff
@@ -339,3 +345,103 @@ def cleanup_old_betatest_users():
         users_to_delete = queryset.filter(is_active=False, profile__modified__lte=delete_cutoff)
         logger.info('Deleting %s users.', users_to_delete.count())
         users_to_delete.delete()
+
+
+class KillZombiesRunner:
+    """
+    Helper class to run `kill_zombies_periodically`
+    """
+
+    def __init__(self):
+        self.enabled: bool = bool(os.environ.get("KILL_ZOMBIES_ENABLED", False))
+        self.region: str = os.environ.get("KILL_ZOMBIES_REGION", "gra")
+        self.recipient: str = os.environ.get(
+            "KILL_ZOMBIES_WARNING_RECIPIENT", "ops@opencraft.com"
+        )
+        self.threshold: int = int(os.environ.get("KILL_ZOMBIES_THRESHOLD", 10))
+
+    def send_email(self, subject: str, body: str) -> int:
+        """
+        Utility method for sending emails
+        """
+        email = EmailMessage(
+            subject, body, settings.DEFAULT_FROM_EMAIL, [self.recipient]
+        )
+        return email.send()
+
+    def trigger_warning(self):
+        """
+        Warns when more than `threshold` VMs will be terminated
+        """
+        subject = "{} Zombie instances over current threshold ({})".format(
+            datetime.now().strftime("%Y-%m-d"), self.threshold
+        )
+        body = (
+            "The number of instances in region {} to be terminated"
+            "is over KILL_ZOMBIES_WARNING_THRESHOLD ({}).\n"
+        ).format(self.region, self.threshold)
+        logging.info(
+            "%s\nSending an email notification to: %s",
+            body, self.recipient
+        )
+        self.send_email(subject, body)
+
+    def handle_dry_run_output(self, stdout: str) -> int:
+        """
+        If there are servers to terminate, a `kill_zombies` dry run will warn:
+        "Would have terminated {} zombies" to stdout.
+        """
+        if "No servers found in region" in stdout:
+            return 0
+        try:
+            return int(stdout.split("Would have terminated")[1].split()[0])
+        except IndexError:
+            logger.info("Received unexpected input from dry run")
+
+
+@db_periodic_task(
+    crontab(
+        minute=KILL_ZOMBIES_CRON_SCHEDULE[0],
+        hour=KILL_ZOMBIES_CRON_SCHEDULE[1],
+        day=KILL_ZOMBIES_CRON_SCHEDULE[2],
+        month=KILL_ZOMBIES_CRON_SCHEDULE[3],
+        day_of_week=KILL_ZOMBIES_CRON_SCHEDULE[4],
+    )
+)
+def kill_zombies_periodically():
+    """
+    Runs the 'kill_zombies' command to terminate servers
+    whose manual termination didn't succeed.
+    Defaults schedule is daily at 12AM
+    """
+    task_runner = KillZombiesRunner()
+    if task_runner.enabled:
```

Comment:
@spokerman12, this would still a cause a DB task to be created, queued and processed by a worker though the task itself might do nothing. So do you think it will be cleaner to do something like what is done [here](https://github.com/open-craft/opencraft/blob/master/backup_swift/tasks.py#L105-L116)?

Reply:
Aha, thanks for pointing this out.

Reply:
After reading a bit, the suggestion seems like the wrong approach as it should be Huey's responsibility to handle the tasks using [revoke](https://huey.readthedocs.io/en/latest/api.html#TaskWrapper.revoke) and pause.
I'll implement the proposal but without a test for toggle=true test because that would be testing Huey.

Reply:
I tested it manually:

```
(opencraft) vagrant@ubuntu-xenial:~/opencraft$ echo $KILL_ZOMBIES_ENABLED
false
...
16:14:02 worker.1              | [2021-03-01 16:14:02,844] INFO:huey.consumer:MainThread:The following commands are available:
16:14:02 worker.1              | + instance.tasks.start_deployment
16:14:02 worker.1              | + instance.tasks.check_deactivation
16:14:02 worker.1              | + instance.tasks.spawn_appserver
16:14:02 worker.1              | + instance.tasks.make_appserver_active
16:14:02 worker.1              | + instance.tasks.shut_down_obsolete_pr_sandboxes
16:14:02 worker.1              | + instance.tasks.terminate_obsolete_appservers_all_instances
16:14:02 worker.1              | + instance.tasks.clean_up
16:14:02 worker.1              | + instance.tasks.delete_old_logs
16:14:02 worker.1              | + instance.tasks.cleanup_old_betatest_users
16:14:02 worker.1              | + registration.tasks.verify_external_domain_configuration

(opencraft) vagrant@ubuntu-xenial:~/opencraft$ echo $KILL_ZOMBIES_ENABLED
true
16:15:50 worker.1              | [2021-03-01 16:15:50,640] INFO:huey.consumer:MainThread:The following commands are available:
16:15:50 worker.1              | + instance.tasks.start_deployment
16:15:50 worker.1              | + instance.tasks.check_deactivation
16:15:50 worker.1              | + instance.tasks.spawn_appserver
16:15:50 worker.1              | + instance.tasks.make_appserver_active
16:15:50 worker.1              | + instance.tasks.shut_down_obsolete_pr_sandboxes
16:15:50 worker.1              | + instance.tasks.terminate_obsolete_appservers_all_instances
16:15:50 worker.1              | + instance.tasks.clean_up
16:15:50 worker.1              | + instance.tasks.delete_old_logs
16:15:50 worker.1              | + instance.tasks.cleanup_old_betatest_users
16:15:50 worker.1              | + instance.tasks.kill_zombies_periodically_task
16:15:50 worker.1              | + registration.tasks.verify_external_domain_configuration

```